libraries/lib-snapping/SnapUtils.cpp

```diff
@@ -0,0 +1,316 @@
+      const auto multiplier =
+         (cfg.tempo / 60.0) * tripletDivisor / cfg.timeSignature.second;
+      return SnapWithMultiplier(value, multiplier, nearest);
+   };
+}
+
+SnapRegistryItemRegistrator beats {
+   Registry::Placement { {}, { Registry::OrderingHint::Begin } },
+   SnapFunctionGroup(
+      "beats", XO("Beats"), true, SnapFunction("bar", XO("Bar"), SnapToBar()),
+      SnapFunction("bar_1_2", XO("1/2"), SnapToBeat(2)),
+      SnapFunction("bar_1_4", XO("1/4"), SnapToBeat(4)),
+      SnapFunction("bar_1_8", XO("1/8"), SnapToBeat(8)),
+      SnapFunction("bar_1_16", XO("1/16"), SnapToBeat(16)),
+      SnapFunction("bar_1_32", XO("1/32"), SnapToBeat(32)),
+      SnapFunction("bar_1_64", XO("1/64"), SnapToBeat(64)),
+      SnapFunction("bar_1_128", XO("1/128"), SnapToBeat(128)))
+};
+
+SnapRegistryItemRegistrator triplets {
+   Registry::Placement { {},
+                         { Registry::OrderingHint::After, "beats" } },
+   SnapFunctionGroup(
+      "triplets", XO("Triplets"), true,
+      SnapFunction("triplet_1_2", XO("1/2 (triplets)"), SnapToTriplets(2)),
+      SnapFunction("triplet_1_4", XO("1/4 (triplets)"), SnapToTriplets(4)),
+      SnapFunction("triplet_1_8", XO("1/8 (triplets)"), SnapToTriplets(8)),
+      SnapFunction("triplet_1_16", XO("1/16 (triplets)"), SnapToTriplets(16)),
+      SnapFunction("triplet_1_32", XO("1/32 (triplets)"), SnapToTriplets(32)),
+      SnapFunction("triplet_1_64", XO("1/64 (triplets)"), SnapToTriplets(64)),
+      SnapFunction(
+         "triplet_1_128", XO("1/128 (triplets)"), SnapToTriplets(128)))
+};
+
+SnapRegistryItemRegistrator secondsAndSamples {
+   Registry::Placement { {},
+                         { Registry::OrderingHint::After, "triplets" } },
+   SnapFunctionGroup(
+      "time", XO("Seconds && samples"), false,
+      SnapFunction("seconds", XO("Seconds"), SnapToTime(1.0)),
+      SnapFunction("deciseconds", XO("Deciseconds"), SnapToTime(10.0)),
+      SnapFunction("centiseconds", XO("Centiseconds"), SnapToTime(100.0)),
+      SnapFunction("milliseconds", XO("Milliseconds"), SnapToTime(1000.0)),
+      SnapFunction("samples", XO("Samples"), SnapToSamples))
+};
+
+SnapRegistryItemRegistrator videoFrames {
+   Registry::Placement { {},
+                         { Registry::OrderingHint::After, "time" } },
+   SnapFunctionGroup(
+      "video", XO("Video frames"), false,
+      SnapFunction(
+         "film_24_fps", XO("Film frames (24 fps)"), SnapToFrames(24.0)),
+      SnapFunction(
+         "film_23.97_fps", XO("NTSC frames (23.97 fps)"),
+         SnapToFrames(24.0 / 1.001)),
+      SnapFunction("film_25_fps", XO("PAL frames"), SnapToFrames(25.0)))
+};
+
+SnapRegistryItemRegistrator cdFrames {
+   Registry::Placement { {},
+                         { Registry::OrderingHint::After, "video" } },
+   SnapFunctionGroup(
+      "cd", XO("CD frames"), false,
+      SnapFunction("cd_75_fps", XO("CDDA frames (75 fps)"), SnapToFrames(75.0)))
+};
```

Comment:
That you for figuring out and reapplying the Registry system.

How would you rate your first experience?

- [ ] 😑 
- [ ] 🙄 
- [ ] 🥱 
- [ ] 🤯 
- [ ] 🤩 
- [ ] 🤮